### PR TITLE
etcdserver: configure "expensive" requests duration

### PR DESCRIPTION
**Pull request by roytman**
When a unary request takes more than predefined duration, this request
is defined as "expensive" and a warning is printed. The expensive request
duration is hard-coded to 300 ms. It can be not enough for example
for transactions with a lot of operations. The warnings just blow up
the log files and reduce throughput.

This fix allows user to configure the "expensive" request duration.

Signed-off-by: Alexey Roytman <roytman@il.ibm.com>

### CODE REVIEW

File: server/etcdmain/config.go [hunk 1]
```diff
@@ -286,6 +286,7 @@ func newConfig() *config {
 	fs.DurationVar(&cfg.ec.ExperimentalWatchProgressNotifyInterval, "experimental-watch-progress-notify-interval", cfg.ec.ExperimentalWatchProgressNotifyInterval, "Duration of periodic watch progress notifications.")
 	fs.DurationVar(&cfg.ec.ExperimentalDowngradeCheckTime, "experimental-downgrade-check-time", cfg.ec.ExperimentalDowngradeCheckTime, "Duration of time between two downgrade status check.")
 	fs.DurationVar(&cfg.ec.ExperimentalWarningApplyDuration, "experimental-warning-apply-duration", cfg.ec.ExperimentalWarningApplyDuration, "Time duration after which a warning is generated if request takes more time.")
+	fs.DurationVar(&cfg.ec.ExperimentalWarningUnaryRequestDuration, "experimental-warning-unary-request-duration", cfg.ec.ExperimentalWarningUnaryRequestDuration, "Time duration after which a warning is generated if a unary request takes more time.")
```
Comment (lilic) [inline server/etcdmain/config.go:289]: Lets add a mention of the default value here as well. Maybe we can print the default value from a variable or this way:

```suggestion
	fs.DurationVar(&cfg.ec.ExperimentalWarningUnaryRequestDuration, "experimental-warning-unary-request-duration", cfg.ec.ExperimentalWarningUnaryRequestDuration, "Time duration after which a warning is generated if a unary request takes more time. Default value is 300ms.")
```

**Review by lilic — COMMENTED** [review]
Thanks! This makes sense to me to be configurable.

@ptabor @hexfusion @serathius @gyuho what are your thoughts here?

File: server/etcdmain/config.go (hunk 1 above)
Comment (roytman) [reply server/etcdmain/config.go:289]: Hi @lilic , thanks for the review. 
No problem, I can add the default value to the explanation, however, the flag help message is comming from the `help.go` file and it does contain the default value.
```  
--experimental-warning-unary-request-duration '300ms'
    Warning is generated if a unary request takes  more than this duration.
```
Futhermore, most of other flag arguments in  `config.go` do not mention their  default values either.
What do you think ?

**Review by spzala — APPROVED** [review]
It's good to let users define what's expensive to them and set the conf value accordingly. LGTM. Thanks @roytman

File: server/etcdmain/help.go [hunk 2]
```diff
@@ -246,6 +246,8 @@ Experimental feature:
     Enable the write transaction to use a shared buffer in its readonly check operations.
   --experimental-bootstrap-defrag-threshold-megabytes
     Enable the defrag during etcd server bootstrap on condition that it will free at least the provided threshold of disk space. Needs to be set to non-zero value to take effect.
+  --experimental-warning-unary-request-duration '300ms'
+    Warning is generated if a unary request takes  more than this duration.
```
Comment (lilic) [inline server/etcdmain/help.go:250]:
```suggestion
    Set time duration after which a warning is generated if a unary request takes more than this duration.
```

**Review by lilic — APPROVED** [review]
Thanks, one suggestion. Otherwise lgtm

File: server/etcdmain/config.go (hunk 1 above)
Comment (lilic) [reply server/etcdmain/config.go:289]: Yes sounds good to me! Added one suggestion for that.

File: server/etcdmain/help.go (hunk 2 above)
Comment (roytman) [reply server/etcdmain/help.go:250]: Thanks, updated

Comment (roytman) [reply server/etcdmain/help.go]: Can you approve the CI process ?

Comment (lilic) [reply server/etcdmain/help.go]: @hexfusion do you mind approving, as reviewer I don't have ability to approve these.

**Review by serathius — APPROVED** [review]
LGTM, we could consider adding an e2e tests that sets this flag to very low value and checks for log present. Normal logs are not as important but those performance warnings might be very important for some users. Might be in separate PR.

File: server/etcdmain/help.go (hunk 2 above)
Comment (spzala) [reply server/etcdmain/help.go]: Approved!

**Comment by spzala** [comment]
Test failure not related.

**Comment by spzala** [comment]
Opened https://github.com/etcd-io/etcd/issues/13264 to address comment from @serathius

**Pull request merged**